In [ ]:
import tensorflow as tf
import keras
import numpy as np
import cv2
import glob

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Define some parameters for the loader

IMG_SIZE = 256
MAX_SEQ_LENGTH = 18
NUM_FEATURES = 2048

In [ ]:
np.random.seed(4)
tf.random.set_seed(4)

# Feature extraction

In [ ]:
# def build_feature_extractor():
#     InceptionV3_model = tf.keras.applications.InceptionV3(
#         weights="imagenet",
#         include_top=False,
#         pooling="avg",
#         input_shape=(IMG_SIZE, IMG_SIZE, 3),
#     )

#     inputs = keras.Input((IMG_SIZE, IMG_SIZE, 3))

#     # preprocess the input
#     preprocess_input = keras.applications.inception_v3.preprocess_input(inputs)

#     # extract the features from the preprocessed input
#     outputs = InceptionV3_model(preprocess_input)

#     # form the final model 
#     myModel = keras.Model(inputs, outputs, name="feature_extractor")
    
#     return myModel


# feature_extractor = build_feature_extractor()

87910968/87910968 [==============================] - 3s 0us/step


In [ ]:
# data_dir_train = 'C:/Users/yzmod/Downloads/OneDrive_1_12-6-2022/03/03/train'
# data_dir_test = 'C:/Users/yzmod/Downloads/OneDrive_1_12-6-2022/03/03/test'

In [ ]:
# path1 = "/content/drive/MyDrive/01/train"
# path2 = "/content/drive/MyDrive/02/train"


In [ ]:
# def prepare_all_samples(root_dir):
    
#     allSamplesFeatures = []
#     labels = np.array([])
#     cv_img = []
    
#     for class_folder in glob.glob(root_dir + "/*"):
#         class_label = int("0127")
#         for folder in glob.glob(class_folder + "/*"):
#             for img in glob.glob(folder + "/*.jpg"):
#                 n = cv2.imread(img)
#                 cv_img.append(n)
#             samplesFeatures = feature_extractor.predict(np.array(cv_img), verbose=0)
#             labels = np.append(labels, class_label)
#             cv_img = []
#             allSamplesFeatures.append(samplesFeatures.squeeze())

#     return np.array(allSamplesFeatures), labels

In [ ]:
# def prepare_all_samples1(root_dir):
    
#     allSamplesFeatures = []
#     labels = np.array([])
#     cv_img = []
    
#     for class_folder in glob.glob(root_dir + "/*"):
#       class_label = int("0160")
#         for folder in glob.glob(class_folder + "/*"):
#             for img in glob.glob(folder + "/*.jpg"):
#                 n = cv2.imread(img)
#                 cv_img.append(n)
#             samplesFeatures = feature_extractor.predict(np.array(cv_img), verbose=0)
#             labels = np.append(labels, class_label)
#             cv_img = []
#             allSamplesFeatures.append(samplesFeatures.squeeze())

#     return np.array(allSamplesFeatures), labels

In [ ]:
#signer01_0127, signer01_0127_label = prepare_all_samples(path1)

In [ ]:
x = signer01_0127[0][14]
x.shape

(2048,)

AttributeError: ignored

In [ ]:
#signer01_0127=np.stack(list(signer01_0127))
#signer01_0127=signer01_0127.reshape((15,2048))

#signer01_0127.shape

(15, 2048)

In [ ]:
#signer01_0127 = np.vstack([signer01_0127,x])


In [ ]:
#signer01_0127.shape

(18, 2048)

In [ ]:
#np.save('signer01_0127_numpy.npy',signer01_0127)
#np.save('signer01_0127_label_numpy.npy',signer01_0127_label)

In [ ]:
#signer02_0160, signer02_0160_label = prepare_all_samples1(path2)

In [ ]:
#x = signer02_0160[9]
#x.shape

(2048,)

In [ ]:
#signer02_0160 = np.stack(list(signer02_0160))
#signer02_0160 = signer02_0160.reshape((10,2048))
#print(signer02_0160.shape)
#print(signer02_0160_label)

(10, 2048)
[160.]


In [ ]:
#for i in range(7):
  #signer02_0160 =np.vstack([signer02_0160,x])
#signer02_0160.shape


(18, 2048)

In [ ]:
#np.save('signer02_0160_numpy.npy',signer02_0160)
#np.save('signer02_0160_label_numpy.npy',signer02_0160_label)

**After checking there is 1 sample in train set of 01, 02 signer that have different sequence length so we decided to drop it and the loop below drops that sample**

In [ ]:
# for i in range(len(train_data)):
#     if train_data[i].shape != (18,2048):
#         print(train_data[i].shape)
#         print(i)
#         train_data = np.delete(train_data, i)
#         train_labels = np.delete(train_labels, i)

**To resolve the shape from ragged to desired shape we need do the method below**

In [ ]:
# train_data = np.stack(list(train_data))

In [ ]:
# test_data, test_labels = prepare_all_samples(data_dir_test)

## Saving extracted values

In [ ]:
# np.save('train_data_03_numpy.npy', train_data)

# np.save('train_labels_03_numpy.npy', train_labels)

In [ ]:
# np.save('test_data_03_numpy.npy', test_data)

# np.save('test_labels_03_numpy.npy', test_labels)

## Loading extracted values

### 01 Signer

In [ ]:
# retrieving train data from file.

train_data01 = np.load('train_data_01_numpy.npy')
print("shape of train data: ", train_data01.shape)

# retrieving labels from file.

train_labels_01 = np.load('train_labels_01_numpy.npy')
print("shape of train labels: ", train_labels_01.shape)

shape of train data:  (3505, 18, 2048)
shape of train labels:  (3505,)


In [ ]:
# retrieving test data from file.
test_data01 = np.load('test_data_01_numpy.npy')
print("shape of test data: ", test_data01.shape)

# retrieving labels from file.

test_labels_01 = np.load('test_labels_01_numpy.npy')
print("shape of test labels: ", test_labels_01.shape)

shape of test data:  (800, 18, 2048)
shape of test labels:  (800,)


### 02 Signer

In [ ]:
# retrieving train data from file.

train_data02 = np.load('train_data_02_numpy.npy')
print("shape of train data: ", train_data02.shape)

# retrieving labels from file.

train_labels_02 = np.load('train_labels_02_numpy.npy')
print("shape of train labels: ", train_labels_02.shape)

shape of train data:  (3565, 18, 2048)
shape of train labels:  (3565,)


In [ ]:
# retrieving test data from file.
test_data02 = np.load('test_data_02_numpy.npy')
print("shape of test data: ", test_data02.shape)

# retrieving labels from file.

test_labels_02 = np.load('test_labels_02_numpy.npy')
print("shape of test labels: ", test_labels_02.shape)

shape of test data:  (800, 18, 2048)
shape of test labels:  (800,)


### 03 Signer

In [ ]:
# retrieving train data from file.

train_data03 = np.load('train_data_03_numpy.npy')
print("shape of train data: ", train_data03.shape)

# retrieving labels from file.

train_labels_03 = np.load('train_labels_03_numpy.npy')
print("shape of train labels: ", train_labels_03.shape)

shape of train data:  (3546, 18, 2048)
shape of train labels:  (3546,)


In [ ]:
# retrieving test data from file.
test_data03 = np.load('test_data_03_numpy.npy')
print("shape of test data: ", test_data03.shape)

# retrieving labels from file.

test_labels_03 = np.load('test_labels_03_numpy.npy')
print("shape of test labels: ", test_labels_03.shape)

shape of test data:  (800, 18, 2048)
shape of test labels:  (800,)


In [ ]:
import pandas as pd

df = pd.read_excel("C:/Users/yzmod/Downloads/KARSL-100_Labels.xlsx")

display(df.head())

SignID Sign-Arabic        Sign-English
0      71   هيكل عظمي            Skeleton
1      72        جمجة               skull
2      73   عمود فقري            Backbone
3      74    قفص صدري               Chest
4      75  جهاز تنفسي  Respiratory device

In [ ]:
label_processor = keras.layers.StringLookup(
    num_oov_indices=0, vocabulary=np.unique(df["Sign-English"])
)
print(label_processor.get_vocabulary())
class_vocab = label_processor.get_vocabulary()

['Aids', 'Appendix', 'Backbone', 'Cerebral paralysis', 'Chest', 'Face', 'Heart failure', 'Ins and Outs', 'Large intestine', 'Loss of hair', 'Respiratory device', 'Skeleton', 'Small intestine', 'Trachea', 'acupuncture', 'addiction', 'adhesive tape', 'allergy', 'analysis', 'analysis laboratory', 'anesthetist', 'bacterium', 'blood analysis', 'blood pressure', 'burning', 'cancer', 'capsule', 'close', 'cold', 'colic', 'constipation', 'descend', 'diabetes', 'diarrhea', 'digestive system', 'drink', 'dropper', "drug's", 'eat', 'epidemic', 'fever', 'first aid', 'five senses', 'gauze', 'headache', 'healthy', 'hear', 'hearing hindrance', 'heart', 'heart attack', 'hemiplegia', 'hindrance', 'hospital', 'immunity', 'infection', 'inflammation', 'inhale', 'inoculation', 'itch', 'laser ray', 'liquid medicine', 'liver', 'lungs', 'medicine', 'mentality hindrance', 'microbe', 'mongoloid', 'monthly circulation', 'muscle', 'nerve', 'nervous system', 'ointment', 'oneness', 'open', 'pain', 'pancreas', 'pharma

In [ ]:
tmp = zip(df['SignID'], df['Sign-English'])
mapping = dict()
for i, j in tmp:
    mapping[i] = j

def map_labels(labels):
    mapped_labels = []
    for i in range(len(labels)):
        mapped_labels.append(mapping[labels[i]])
    return np.array(mapped_labels)

NameError: ignored

## Split train data

Need labels values between range [0, 100)

In [ ]:
train_labels_01 = train_labels_01 - 71
test_labels_01 = test_labels_01 - 71
train_labels_02 = train_labels_02 - 71
test_labels_02 = test_labels_02 - 71
train_labels_03 = train_labels_03 - 71
test_labels_03 = test_labels_03 - 71

In [ ]:
from sklearn.model_selection import train_test_split

train_data01, val_data01, train_labels_01, val_labels_01 = train_test_split(train_data01, train_labels_01, test_size=0.3, random_state=4, stratify=train_labels_01)
train_data02, val_data02, train_labels_02, val_labels_02 = train_test_split(train_data02, train_labels_02, test_size=0.3, random_state=4, stratify=train_labels_02)
train_data03, val_data03, train_labels_03, val_labels_03 = train_test_split(train_data03, train_labels_03, test_size=0.3, random_state=4, stratify=train_labels_03)

# Training Dependent models

In [ ]:
from sklearn.metrics import f1_score, classification_report, confusion_matrix, matthews_corrcoef
from imblearn.metrics import geometric_mean_score

## Signer 1 Dependent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Dep01"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.2)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)       

In [ ]:
# start the training
history = seq_model.fit(
    train_data01,
    train_labels_01,
    batch_size=32,
    validation_data=(val_data01, val_labels_01),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
77/77 [==============================] - ETA: 0s - loss: 4.6465 - accuracy: 0.0416
Epoch 1: val_loss improved from inf to 4.55682, saving model to C:/Users/yzmod/Downloads/Model\Dep01
77/77 [==============================] - 16s 118ms/step - loss: 4.6465 - accuracy: 0.0416 - val_loss: 4.5568 - val_accuracy: 0.1150
Epoch 2/15
77/77 [==============================] - ETA: 0s - loss: 4.4535 - accuracy: 0.1447
Epoch 2: val_loss improved from 4.55682 to 4.29313, saving model to C:/Users/yzmod/Downloads/Model\Dep01
77/77 [==============================] - 7s 91ms/step - loss: 4.4535 - accuracy: 0.1447 - val_loss: 4.2931 - val_accuracy: 0.2662
Epoch 3/15
77/77 [==============================] - ETA: 0s - loss: 4.1373 - accuracy: 0.2638
Epoch 3: val_loss improved from 4.29313 to 3.92993, saving model to C:/Users/yzmod/Downloads/Model\Dep01
77/77 [==============================] - 7s 92ms/step - loss: 4.1373 - accuracy: 0.2638 - val_loss: 3.9299 - val_accuracy: 0.2757
Epoch 4/15
77/7

In [ ]:
import matplotlib.pyplot as plt

print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data01, test_labels_01)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 1s 40ms/step - loss: 0.7840 - accuracy: 0.9262
Test accuracy: 92.62%


In [ ]:
y_pred = seq_model.predict(test_data01, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_01, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_01, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_01, y_pred))
print(classification_report(test_labels_01, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.914
The MCC score is: 0.926
The GMean score is: 0.000
                      precision    recall  f1-score   support

            Skeleton       1.00      1.00      1.00         8
               skull       1.00      1.00      1.00         8
            Backbone       1.00      1.00      1.00         8
               Chest       1.00      1.00      1.00         8
  Respiratory device       1.00      1.00      1.00         8
             Trachea       1.00      1.00      1.00         8
               lungs       1.00      0.88      0.93         8
        Ins and Outs       0.50      0.12      0.20         8
    digestive system       1.00      1.00      1.00         8
                Face       1.00      1.00      1.00         8
             pharynx       1.00      1.00      1.00         8
               liver       1.00      1.00      1.00         8
            pancreas       0.89      1.00      0.94         8
     Small intestine       1.00      1.00      1.00       

C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Signer 2 Dependent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Dep02"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.2)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)       

In [ ]:
# start the training
history = seq_model.fit(
    train_data02,
    train_labels_02,
    batch_size=32,
    validation_data=(val_data02, val_labels_02),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
78/78 [==============================] - ETA: 0s - loss: 4.6599 - accuracy: 0.0281
Epoch 1: val_loss improved from inf to 4.56815, saving model to C:/Users/yzmod/Downloads/Model\Dep02
78/78 [==============================] - 16s 116ms/step - loss: 4.6599 - accuracy: 0.0281 - val_loss: 4.5682 - val_accuracy: 0.0841
Epoch 2/15
78/78 [==============================] - ETA: 0s - loss: 4.4624 - accuracy: 0.1299
Epoch 2: val_loss improved from 4.56815 to 4.30204, saving model to C:/Users/yzmod/Downloads/Model\Dep02
78/78 [==============================] - 8s 107ms/step - loss: 4.4624 - accuracy: 0.1299 - val_loss: 4.3020 - val_accuracy: 0.2178
Epoch 3/15
78/78 [==============================] - ETA: 0s - loss: 4.1128 - accuracy: 0.2473
Epoch 3: val_loss improved from 4.30204 to 3.83410, saving model to C:/Users/yzmod/Downloads/Model\Dep02
78/78 [==============================] - 8s 100ms/step - loss: 4.1128 - accuracy: 0.2473 - val_loss: 3.8341 - val_accuracy: 0.3701
Epoch 4/15
78

In [ ]:
print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data02, test_labels_02)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 1s 48ms/step - loss: 0.4532 - accuracy: 0.9775
Test accuracy: 97.75%


In [ ]:
y_pred = seq_model.predict(test_data02, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_02, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_02, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_02, y_pred))
print(classification_report(test_labels_02, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.977
The MCC score is: 0.977
The GMean score is: 0.976
                      precision    recall  f1-score   support

            Skeleton       1.00      1.00      1.00         8
               skull       1.00      0.88      0.93         8
            Backbone       1.00      1.00      1.00         8
               Chest       1.00      1.00      1.00         8
  Respiratory device       1.00      1.00      1.00         8
             Trachea       1.00      1.00      1.00         8
               lungs       1.00      1.00      1.00         8
        Ins and Outs       1.00      1.00      1.00         8
    digestive system       1.00      1.00      1.00         8
                Face       0.89      1.00      0.94         8
             pharynx       1.00      1.00      1.00         8
               liver       0.89      1.00      0.94         8
            pancreas       0.89      1.00      0.94         8
     Small intestine       0.89      1.00      0.94       

## Signer 3 Dependent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Dep03"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.2)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)   

In [ ]:
# start the training
history = seq_model.fit(
    train_data03,
    train_labels_03,
    batch_size=32,
    validation_data=(val_data03, val_labels_03),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
78/78 [==============================] - ETA: 0s - loss: 4.6348 - accuracy: 0.0342
Epoch 1: val_loss improved from inf to 4.52318, saving model to C:/Users/yzmod/Downloads/Model\Dep03
78/78 [==============================] - 16s 122ms/step - loss: 4.6348 - accuracy: 0.0342 - val_loss: 4.5232 - val_accuracy: 0.0818
Epoch 2/15
78/78 [==============================] - ETA: 0s - loss: 4.3998 - accuracy: 0.1398
Epoch 2: val_loss improved from 4.52318 to 4.22835, saving model to C:/Users/yzmod/Downloads/Model\Dep03
78/78 [==============================] - 8s 100ms/step - loss: 4.3998 - accuracy: 0.1398 - val_loss: 4.2284 - val_accuracy: 0.2585
Epoch 3/15
78/78 [==============================] - ETA: 0s - loss: 4.0298 - accuracy: 0.2768
Epoch 3: val_loss improved from 4.22835 to 3.75631, saving model to C:/Users/yzmod/Downloads/Model\Dep03
78/78 [==============================] - 8s 99ms/step - loss: 4.0298 - accuracy: 0.2768 - val_loss: 3.7563 - val_accuracy: 0.4436
Epoch 4/15
78/

In [ ]:
print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data03, test_labels_03)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 1s 42ms/step - loss: 0.4145 - accuracy: 0.9912
Test accuracy: 99.12%


In [ ]:
y_pred = seq_model.predict(test_data03, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_03, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_03, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_03, y_pred))
print(classification_report(test_labels_03, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.991
The MCC score is: 0.991
The GMean score is: 0.990
                      precision    recall  f1-score   support

            Skeleton       1.00      1.00      1.00         8
               skull       1.00      1.00      1.00         8
            Backbone       1.00      1.00      1.00         8
               Chest       1.00      1.00      1.00         8
  Respiratory device       1.00      1.00      1.00         8
             Trachea       1.00      1.00      1.00         8
               lungs       1.00      1.00      1.00         8
        Ins and Outs       1.00      1.00      1.00         8
    digestive system       1.00      1.00      1.00         8
                Face       1.00      1.00      1.00         8
             pharynx       1.00      1.00      1.00         8
               liver       1.00      1.00      1.00         8
            pancreas       1.00      1.00      1.00         8
     Small intestine       0.89      1.00      0.94       

# Training Independent models

## Signer 1 Independent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Indep01"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.4)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)    

In [ ]:
train_data02_03 = np.vstack([train_data02, train_data03])
val_data02_03 = np.vstack([val_data02, val_data03])
train_labels_02_03 = np.append(train_labels_02, train_labels_03)
val_labels_02_03 = np.append(val_labels_02, val_labels_03)

In [ ]:
# start the training
history = seq_model.fit(
    train_data02_03,
    train_labels_02_03,
    batch_size=64,
    validation_data=(val_data02_03, val_labels_02_03),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
78/78 [==============================] - ETA: 0s - loss: 4.6703 - accuracy: 0.0221
Epoch 1: val_loss improved from inf to 4.60208, saving model to C:/Users/yzmod/Downloads/Model\Indep01
78/78 [==============================] - 27s 266ms/step - loss: 4.6703 - accuracy: 0.0221 - val_loss: 4.6021 - val_accuracy: 0.0651
Epoch 2/15
78/78 [==============================] - ETA: 0s - loss: 4.5318 - accuracy: 0.0701
Epoch 2: val_loss improved from 4.60208 to 4.39901, saving model to C:/Users/yzmod/Downloads/Model\Indep01
78/78 [==============================] - 15s 192ms/step - loss: 4.5318 - accuracy: 0.0701 - val_loss: 4.3990 - val_accuracy: 0.1771
Epoch 3/15
78/78 [==============================] - ETA: 0s - loss: 4.2714 - accuracy: 0.1453
Epoch 3: val_loss improved from 4.39901 to 4.02520, saving model to C:/Users/yzmod/Downloads/Model\Indep01
78/78 [==============================] - 15s 196ms/step - loss: 4.2714 - accuracy: 0.1453 - val_loss: 4.0252 - val_accuracy: 0.2741
Epoch

In [ ]:
print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data01, test_labels_01)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 2s 61ms/step - loss: 4.7080 - accuracy: 0.0825
Test accuracy: 8.25%


In [ ]:
y_pred = seq_model.predict(test_data01, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_01, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_01, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_01, y_pred))
print(classification_report(test_labels_01, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.065
The MCC score is: 0.074
The GMean score is: 0.000
                      precision    recall  f1-score   support

            Skeleton       0.00      0.00      0.00         8
               skull       0.12      0.38      0.18         8
            Backbone       0.00      0.00      0.00         8
               Chest       0.00      0.00      0.00         8
  Respiratory device       0.00      0.00      0.00         8
             Trachea       0.00      0.00      0.00         8
               lungs       0.00      0.00      0.00         8
        Ins and Outs       0.00      0.00      0.00         8
    digestive system       0.00      0.00      0.00         8
                Face       0.00      0.00      0.00         8
             pharynx       0.00      0.00      0.00         8
               liver       0.00      0.00      0.00         8
            pancreas       0.00      0.00      0.00         8
     Small intestine       0.00      0.00      0.00       

C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Signer 2 Independent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Indep02"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.4)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)    

In [ ]:
train_data01_03 = np.vstack([train_data01, train_data03])
val_data01_03 = np.vstack([val_data01, val_data03])
train_labels_01_03 = np.append(train_labels_01, train_labels_03)
val_labels_01_03 = np.append(val_labels_01, val_labels_03)

In [ ]:
# start the training
history = seq_model.fit(
    train_data01_03,
    train_labels_01_03,
    batch_size=64,
    validation_data=(val_data01_03, val_labels_01_03),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
78/78 [==============================] - ETA: 0s - loss: 4.6742 - accuracy: 0.0274
Epoch 1: val_loss improved from inf to 4.61160, saving model to C:/Users/yzmod/Downloads/Model\Indep02
78/78 [==============================] - 21s 192ms/step - loss: 4.6742 - accuracy: 0.0274 - val_loss: 4.6116 - val_accuracy: 0.0832
Epoch 2/15
78/78 [==============================] - ETA: 0s - loss: 4.5509 - accuracy: 0.0772
Epoch 2: val_loss improved from 4.61160 to 4.43840, saving model to C:/Users/yzmod/Downloads/Model\Indep02
78/78 [==============================] - 12s 160ms/step - loss: 4.5509 - accuracy: 0.0772 - val_loss: 4.4384 - val_accuracy: 0.1588
Epoch 3/15
78/78 [==============================] - ETA: 0s - loss: 4.3263 - accuracy: 0.1289
Epoch 3: val_loss improved from 4.43840 to 4.13401, saving model to C:/Users/yzmod/Downloads/Model\Indep02
78/78 [==============================] - 13s 169ms/step - loss: 4.3263 - accuracy: 0.1289 - val_loss: 4.1340 - val_accuracy: 0.2353
Epoch

In [ ]:
print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data02, test_labels_02)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 2s 58ms/step - loss: 4.7352 - accuracy: 0.0962
Test accuracy: 9.62%


In [ ]:
y_pred = seq_model.predict(test_data02, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_02, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_02, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_02, y_pred))
print(classification_report(test_labels_02, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.082
The MCC score is: 0.088
The GMean score is: 0.000
                      precision    recall  f1-score   support

            Skeleton       0.62      0.62      0.62         8
               skull       0.33      0.38      0.35         8
            Backbone       1.00      0.38      0.55         8
               Chest       0.00      0.00      0.00         8
  Respiratory device       0.50      0.12      0.20         8
             Trachea       0.00      0.00      0.00         8
               lungs       0.00      0.00      0.00         8
        Ins and Outs       0.50      0.25      0.33         8
    digestive system       0.00      0.00      0.00         8
                Face       0.00      0.00      0.00         8
             pharynx       0.00      0.00      0.00         8
               liver       0.00      0.00      0.00         8
            pancreas       0.00      0.00      0.00         8
     Small intestine       0.00      0.00      0.00       

C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Signer 3 Independent

In [ ]:
frame_features_input = keras.Input((MAX_SEQ_LENGTH, NUM_FEATURES))

filepath = "C:/Users/yzmod/Downloads/Model/Indep03"
x = keras.layers.LSTM(64, return_sequences=True)(frame_features_input)
x = keras.layers.LSTM(64)(x)
x = keras.layers.Dropout(0.2)(x)
x = keras.layers.Dense(256, kernel_regularizer=keras.regularizers.l2(0.001), activation="relu")(x)
output = keras.layers.Dense(len(class_vocab), activation="softmax")(x)

seq_model = keras.Model(frame_features_input, output)

seq_model.compile(loss="sparse_categorical_crossentropy", 
                  optimizer=keras.optimizers.Adam(learning_rate = 1e-4), metrics=["accuracy"])

checkpoint = keras.callbacks.ModelCheckpoint(filepath, save_weights_only=True, save_best_only=True, verbose=1)
earlyStopper = keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)    

In [ ]:
train_data01_02 = np.vstack([train_data01, train_data02])
val_data01_02 = np.vstack([val_data01, val_data02])
train_labels_01_02 = np.append(train_labels_01, train_labels_02)
val_labels_01_02 = np.append(val_labels_01, val_labels_02)

In [ ]:
# start the training
history = seq_model.fit(
    train_data01_02,
    train_labels_01_02,
    batch_size=64,
    validation_data=(val_data01_02, val_labels_01_02),
    epochs=15,
    callbacks=[checkpoint, earlyStopper])

Epoch 1/15
78/78 [==============================] - ETA: 0s - loss: 4.6568 - accuracy: 0.0394
Epoch 1: val_loss improved from inf to 4.56341, saving model to C:/Users/yzmod/Downloads/Model\Indep03
78/78 [==============================] - 33s 332ms/step - loss: 4.6568 - accuracy: 0.0394 - val_loss: 4.5634 - val_accuracy: 0.0881
Epoch 2/15
78/78 [==============================] - ETA: 0s - loss: 4.4428 - accuracy: 0.1035
Epoch 2: val_loss improved from 4.56341 to 4.26128, saving model to C:/Users/yzmod/Downloads/Model\Indep03
78/78 [==============================] - 22s 289ms/step - loss: 4.4428 - accuracy: 0.1035 - val_loss: 4.2613 - val_accuracy: 0.1569
Epoch 3/15
78/78 [==============================] - ETA: 0s - loss: 4.0618 - accuracy: 0.1797
Epoch 3: val_loss improved from 4.26128 to 3.77218, saving model to C:/Users/yzmod/Downloads/Model\Indep03
78/78 [==============================] - 22s 286ms/step - loss: 4.0618 - accuracy: 0.1797 - val_loss: 3.7722 - val_accuracy: 0.2516
Epoch

In [ ]:
print(history.history.keys())
#  "Accuracy"
plt.plot(history.history['acc'])
plt.plot(history.history['val_acc'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()
# "Loss"
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'validation'], loc='upper left')
plt.show()

In [ ]:
# use the best model saved during training process for testing
seq_model.load_weights(filepath)
_, accuracy = seq_model.evaluate(test_data03, test_labels_03)

print(f"Test accuracy: {round(accuracy * 100, 2)}%")

25/25 [==============================] - 3s 106ms/step - loss: 4.9405 - accuracy: 0.0688
Test accuracy: 6.88%


In [ ]:
y_pred = seq_model.predict(test_data03, verbose=0)
y_pred = np.argmax(y_pred, axis=1)

print("The f1 score is: %.3f" % f1_score(test_labels_03, y_pred, average='macro'))
print("The MCC score is: %.3f" % matthews_corrcoef(test_labels_03, y_pred))
print("The GMean score is: %.3f" % geometric_mean_score(test_labels_03, y_pred))
print(classification_report(test_labels_03, y_pred, target_names=df['Sign-English'].unique()))

The f1 score is: 0.059
The MCC score is: 0.060
The GMean score is: 0.000
                      precision    recall  f1-score   support

            Skeleton       0.03      0.12      0.05         8
               skull       0.00      0.00      0.00         8
            Backbone       0.04      0.12      0.06         8
               Chest       0.00      0.00      0.00         8
  Respiratory device       0.00      0.00      0.00         8
             Trachea       0.00      0.00      0.00         8
               lungs       0.00      0.00      0.00         8
        Ins and Outs       0.00      0.00      0.00         8
    digestive system       0.00      0.00      0.00         8
                Face       0.00      0.00      0.00         8
             pharynx       0.00      0.00      0.00         8
               liver       0.00      0.00      0.00         8
            pancreas       0.00      0.00      0.00         8
     Small intestine       0.00      0.00      0.00       

C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yzmod\AppData\Roaming\Python\Python38\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Dependent Results
| Signer | Accuracy |
| --- | --- |
| 01 | 92.62% |
| 02 | 97.75% |
| 03 | 99.12% |

# Independent Results
| Signer | Accuracy |
| --- | --- |
| 01 (trained on 02,03) |  8.25% |
| 02 (trained on 01,03) |  9.62% |
| 03 (trained on 01,02) |  6.88% |